source/slang/slang-check-stmt.cpp

```diff
@@ -327,6 +329,260 @@ namespace Slang
+        {
+            if (opName->text == "++")
+                sideEffectFuncOp = kIROp_Add;
+            else if (opName->text == "--")
+                sideEffectFuncOp = kIROp_Sub;
+            else if (opName->text == "+=")
+                sideEffectFuncOp = kIROp_Add;
+            else if (opName->text == "-=")
+                sideEffectFuncOp = kIROp_Sub;
+            else
+                return;
+        }
+        if (opSideEffectExpr->arguments.getCount())
+        {
+            auto varExpr = as<VarExpr>(opSideEffectExpr->arguments[0]);
+            if (!varExpr)
+                return;
+            if (varExpr->declRef != initialVar)
+            {
+                // If the user writes something like `for (int i = 0; i < 5; j++)`,
+                // it is most likely a bug, so we issue a warning.
+                if (predicateVar == initialVar)
+                    getSink()->diagnose(varExpr, Diagnostics::forLoopSideEffectChangingDifferentVar, initialVar, varExpr->declRef);
+                return;
+            }
+        }
+        else
+            return;
+        if (opSideEffectExpr->arguments.getCount() == 2)
+        {
+            auto stepVal = tryFoldIntegerConstantExpression(opSideEffectExpr->arguments[1], nullptr);
+            if (!stepVal)
+                return;
+            if (auto constantIntVal = as<ConstantIntVal>(stepVal))
+            {
+                stepSize = constantIntVal;
+            }
+        }
+        else
+        {
+            stepSize = m_astBuilder->getIntVal(m_astBuilder->getIntType(), 1);
+        }
+
+        if (predicateVar != initialVar)
+        {
+            if (predicateVar)
+                getSink()->diagnose(stmt->predicateExpression, Diagnostics::forLoopPredicateCheckingDifferentVar, initialVar, predicateVar);
+            return;
+        }
+        if (!stepSize)
+            return;
+        if (stepSize->value > 0)
+        {
+            if (sideEffectFuncOp == kIROp_Add && compareOp == kIROp_Greater ||
+                sideEffectFuncOp == kIROp_Sub && compareOp == kIROp_Less)
+            {
+                getSink()->diagnose(stmt->sideEffectExpression, Diagnostics::forLoopChangingIterationVariableInOppsoiteDirection, initialVar);
+                return;
+            }
+        }
+        else if (stepSize->value < 0)
+        {
+            if (sideEffectFuncOp == kIROp_Add && compareOp == kIROp_Less ||
+                sideEffectFuncOp == kIROp_Sub && compareOp == kIROp_Greater)
+            {
+                getSink()->diagnose(stmt->sideEffectExpression, Diagnostics::forLoopChangingIterationVariableInOppsoiteDirection, initialVar);
+                return;
+            }
+        }
+        else
+        {
+            getSink()->diagnose(stmt->sideEffectExpression, Diagnostics::forLoopNotModifyingIterationVariable, initialVar);
+            return;
+        }
+
+        if (!initialLitVal || !finalVal)
```

Comment:
Don't we need to worry about side effects in the body?

Reply:
Yeah, for that we will need to track the use of the induction variable when we check the body. Will implement it next.

Reply:
This is now implemented.

Reply:
I didn't see a test that checked for break in the body?

Reply:
Break is fine because it only makes the loop terminate earlier. As long as the result we derive is conservative things are good.